### DISCUSSION TITLE
Export logs programatically (without presenting console)

### DISCUSSION THREAD

**Discussion by darrarski** [kean/Pulse#176] [Q&A]
Q: Is there a way to export logs programmatically, without presenting the PulseUI console to the user?

We still want to have the console in the app, because it's great and useful for more advanced users and our development team. However, for casual users, we need something easier to use, like a single button in the app, that on tap brings the share sheet with the logs file ready for export.

**Comment by kean** [comment]
Hey, yes, you can export the store using [`LoggerStore/copy`](https://kean-docs.github.io/pulse/documentation/pulse/loggerstore/copy(to:predicate:)), which creates a `.pulse` file. PulseUI supports other export options: HTML, Plain Text PDF. These are not currently publicly available, but it would make a lot of sense to open them up to everyone.

**Reply by darrarski** [reply]
Hi @kean! Thanks for the support. 

I'm experiencing issues when exporting the logs with [`LoggerStore/copy`](https://kean-docs.github.io/pulse/documentation/pulse/loggerstore/copy(to:predicate:)) function. Here is what I do:

```swift
let store = LoggerStore.shared
let dateFormatter = DateFormatter()
dateFormatter.dateFormat = "yyyy-MM-dd-HH-mm-ss"
let filename = "app-logs-\(dateFormatter.string(from: Date()))"
let url = URL.temporaryDirectory
  .appending(component: filename)
  .appendingPathExtension("pulse")
try store.copy(to: url)
```

I can see the file is created at provided URL. According to documentation, it's a `SQLite` file, but I'm struggling with reading its contents. All I can see are the following tables:

![Screenshot 2023-03-27 at 14 43 38](https://user-images.githubusercontent.com/1384684/227943895-e08be45e-c89e-418d-bec8-0f9a08be89ae.png)

but no logs. I can open the file in Pulse Pro, and my app's logs are there. However, the labels are missing:

![Screenshot 2023-03-27 at 14 47 00](https://user-images.githubusercontent.com/1384684/227944524-97577636-aabf-42a1-910c-7634696bda17.png)

Moreover, I can't access log details. When tapping on the following icon, the Pulse Pro app is crashing:

![Screenshot 2023-03-27 at 14 50 20](https://user-images.githubusercontent.com/1384684/227944939-6b084c77-6029-4853-8b03-81edc532d7cb.png)

I can browse the logs with any issues using in-app PulseUI console, and all the informations are present there, including labels and details.

I'm using the latest versions of Pulse (3.5.4) and Pulse Pro (2.0.1).

---

By the way, there is a typo in `Pulse.LoggerStore.Error.fileAleadyExists`: "aleady" → "already".

**Reply by kean** [reply]
> According to documentation, it's a SQLite file, but I'm struggling with reading its contents.

I use SQLite as a document format. The logs (Core Data) database is just one of the files in this SQLite database. It isn't optimal if you want to view the Core Data database directly.

Pulse actually supports two document formats: a single document (SQLite-based) or a package (a special kind of folder on Apple file systems). I think it's worth making the second option public. It does use a bit more space and it makes it harder to share it as it's a folder, and not a single file.

> Moreover, I can't access log details. When tapping on the following icon, the Pulse Pro app is crashing:

Pulse Pro 2 no longer (fully) supports Pulse 3 because it had some significant changes to the store format. I'm currently working on a new Pulse app for macOS (Pulse Pro replacement) written in SwiftUI. The first [alpha version](https://github.com/kean/Pulse/releases/tag/3.6.0-alpha.1) shipped just yesterday. It fully supports Pulse 3 and has a lot of new features from PulseUI 3, such as the new [search](https://kean.blog/post/pulse-search).

> By the way, there is a typo in Pulse.LoggerStore.Error.fileAleadyExists: "aleady" → "already".
 Write

Thanks, will fix!
